src/autopas/selectors/tuningStrategy/PredictiveTuning.h

```diff
@@ -0,0 +1,439 @@
+                                                                       (traversal1.first - traversal2.first) *
+                                                                       (_iterations - traversal1.first);
+  }
+}
+
+void PredictiveTuning::reselectOptimalSearchSpace() {
+  // This is placed here, because there are no unnecessary removals
+  for (auto &configuration : _optimalSearchSpace) {
+    _configurationPredictions.erase(configuration);
+    _validSearchSpace.erase(configuration);
+  }
+
+  _optimalSearchSpace.clear();
+
+  if (_validSearchSpace.size() == 1) {
+    _optimalSearchSpace = _validSearchSpace;
+    _currentConfig = _optimalSearchSpace.begin();
+    return;
+  }
+
+  // checks if there are any configurations left that can be tested
+  if (_configurationPredictions.empty()) {
+    autopas::utils::ExceptionHandler::exception("Predictive Tuning: No valid configuration could be found");
+  }
+
+  const auto optimum = std::min_element(_configurationPredictions.begin(), _configurationPredictions.end(),
+                                        [](std::pair<Configuration, size_t> a,
+                                           std::pair<Configuration, size_t> b) -> bool { return a.second < b.second; });
+
+  if (_validSearchSpace.count(optimum->first) == 0) {
+    autopas::utils::ExceptionHandler::exception("Predictive Tuning: No valid optimal configuration could be found");
+  }
+
+  _optimalSearchSpace.emplace(optimum->first);
+
+  // selects configurations that are near the optimal prediction
+  for (auto &configuration : _validSearchSpace) {
+    // Adds configurations that are within the _relativeOptimumRange
+    if ((float)_configurationPredictions[configuration] / optimum->second <= _relativeOptimumRange) {
+      _optimalSearchSpace.emplace(configuration);
+      _longNotTestedSearchSpace.erase(configuration);
+    }
+  }
+
+  // sanity check
+  if (_optimalSearchSpace.empty()) {
+    autopas::utils::ExceptionHandler::exception("PredicitveTuning: No possible configuration prediction found!");
+  }
+
+  _currentConfig = _optimalSearchSpace.begin();
+}
+
+bool PredictiveTuning::tune(bool valid) {
+  if (not valid) {
+    _validConfigurationFound = true;
+  }
+
+  // repeat as long as traversals are not applicable or we run out of configs
+  ++_currentConfig;
+
+  if (_currentConfig == _searchSpace.end() || _currentConfig == _optimalSearchSpace.end()) {
+    if (_validConfigurationFound) {
+      selectOptimalConfiguration();
+      if (_longNotTestedSearchSpace.empty()) {
+        _currentConfig = _optimalConfig;
+        _tuningIterationsCounter++;
+        return false;
+      } else {
+        _currentConfig = _longNotTestedSearchSpace.begin();
+        return true;
+      }
+    } else {
+      reselectOptimalSearchSpace();
+    }
+  } else if (_currentConfig == _longNotTestedSearchSpace.end()) {
+    _currentConfig = _optimalConfig;
+    _tuningIterationsCounter++;
+    return false;
+  }
```

Comment:
I might misinterpret this but as far as I see, you do not consider the case that a long untested configuration might be actually the best, do you? This is because you call `selectOptimalConfiguration()` before finishing all tests. So you would need to move the call to `selectOptimalConfiguration()` inside of `if(_longNotTestedSearchSpace.empty())` AND this `else if` block.

Reply:
I already changed that.